cpp/src/arrow/csv/parser.cc

```diff
@@ -368,6 +394,9 @@ Status BlockParser::DoParseSpecialized(const char* start, uint32_t size, bool is
     RETURN_NOT_OK(ParseChunk<SpecializedOptions>(&values_writer, &parsed_writer, data,
                                                  data_end, is_final, rows_in_chunk, &data,
                                                  &finished_parsing));
+    if (num_cols_ == -1) {
```

Comment:
The `num_cols_ == -1` check nested in a `num_cols_ == -1` check is a bit confusing.

Reply:
I'm not sure how to make it less confusing, though. I didn't want to pass an extra out parameter to the subroutine.